ngraph_bridge/ngraph_executor.cc

```diff
@@ -0,0 +1,553 @@
+  // Use tensorflow input tensors to get input_shapes, static_input_map
+  // and compute the signature
+  for (int i = 0; i < tf_input_tensors.size(); i++) {
+    const Tensor& input_tensor = tf_input_tensors[i];
+    input_shapes.push_back(input_tensor.shape());
+    for (const auto& x : input_tensor.shape()) {
+      signature_ss << x.size << ",";
+    }
+    signature_ss << ";";
+  }
+
+  signature_ss << "/";
+
+  static_input_map.resize(tf_input_tensors.size());
+  for (int i = 0; i < tf_input_tensors.size(); i++) {
+    const Tensor& input_tensor = tf_input_tensors[i];
+    if (m_input_is_static[i]) {
+      static_input_map[i] = &input_tensor;
+      TF_RETURN_IF_ERROR(TensorToStream(signature_ss, input_tensor));
+      signature_ss << ";";
+    }
+  }
+  return Status::OK();
+}
+
+//---------------------------------------------------------------------------
+//  NGraphExecutor::GetNgExecutable
+//---------------------------------------------------------------------------
+Status NGraphExecutor::GetNgExecutable(
+    const std::vector<Tensor>& tf_input_tensors,
+    std::shared_ptr<ngraph::runtime::Executable>& ng_exec, bool& cache_hit) {
+  NGRAPH_VLOG(4) << "GetNgExecutable: Got backend of type: "
+                 << m_op_backend_name;
+
+  // Get the backend. Note that the backend may not be available
+  // so that's a programmng error.
+  ng::runtime::Backend* op_backend;
+  try {
+    op_backend = BackendManager::GetBackend(m_op_backend_name);
+  } catch (...) {
+    return errors::Internal("Backend not available: ", m_op_backend_name);
+  }
+
+  lock_guard<mutex> lock(m_mutex);
+
+  std::stringstream signature_ss;
+  std::vector<TensorShape> input_shapes;
+  std::vector<const Tensor*> static_input_map;
+  TF_RETURN_IF_ERROR(ComputeSignature(tf_input_tensors, input_shapes,
+                                      static_input_map, signature_ss));
+  string signature;
+  signature = signature_ss.str();
+
+  NGRAPH_VLOG(5) << "Computed signature: " << signature;
+
+  cache_hit = false;
+  auto it = m_ng_exec_map.find(signature);
+
+  // Translate the TensorFlow graph to nGraph.
+  if (it == m_ng_exec_map.end()) {
+    // Measure the current total memory usage
+    long vm, rss, vm0, rss0;
+    MemoryProfile(vm0, rss0);
+
+    NGRAPH_VLOG(1) << "Compilation cache miss: " << m_node_name;
+
+    std::shared_ptr<ngraph::Function> ng_function;
+    std::shared_ptr<ngraph::runtime::Executable> evicted_ng_exec;
+
+    if (!m_do_aot) {
+      TF_RETURN_IF_ERROR(Builder::TranslateGraph(input_shapes, static_input_map,
+                                                 m_graph.get(), ng_function));
+      ng_function->set_friendly_name(m_node_name);
+    } else {
+      auto itr = m_aot_functions.find(signature);
+      if (itr == m_aot_functions.end()) {
+        return errors::Internal(
+            "Expected to find AOT precompiled ng function of signature: ",
+            signature);
+      }
+      ng_function = ng::deserialize(itr->second);
```

Comment:
We are deserializing the string here. We do not need to deserialize (and deserializing will break networks like ssd).

In impl, we do not cache the ng_function at all: https://github.com/tensorflow/ngraph-bridge/blob/master/ngraph_bridge/ngraph_encapsulate_impl.cc#L151

we just have a map of exec->string (where string is the serialized function) rather than exec->function

Reply:
Please fix this once my PR is in.